File: 20_Valid Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let map = new Map();
    map.set(")", "(");
    map.set("}", "{");
    map.set("]", "[");
    
    const size = s.length;
    let stack = [];
    
    for(let i = 0; i < size; ++i) {
        let c = s.charAt(i);
        
        if(map.has(c)) {
            let top = stack.length === 0 ? "#" : stack.pop();
            if (top != map.get(c)) return false;
        } else stack.push(c);
    }
    
    return stack.length === 0 ? true : false;
};

File: 20_Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        std::unordered_map<char, char> map = {
            {')', '('}, {'}', '{'}, {']', '['}  
        };
        
        std::stack<char> st;
        
        for(const auto& str : s) {
            if(map.find(str) != map.end()) {
                if(st.size() > 0) {
                    auto top = st.top();
                    st.pop();
                
                    if(top != map[str]) {
                        return false;
                    }    
                } else {
                    return false;
                }
                
            } else {
                st.push(str);    
            }
        }
        
        return st.empty();
    }
};